egld-esdt-swap/src/lib.rs

```diff
@@ -82,24 +105,75 @@ pub trait EgldEsdtSwap {
 
     // private
 
-    fn data_or_empty(&self, to: &Address, data: &'static [u8]) -> &[u8] {
-        if self.blockchain().is_smart_contract(to) {
-            &[]
-        } else {
-            data
-        }
+    fn needs_execution(&self, caller: &Address, function: &BoxedBytes) -> bool {
+        self.blockchain().is_smart_contract(caller) && !function.is_empty()
+    }
+
+    fn transfer_and_execute_via_async(
+        &self,
+        caller: &Address,
+        token_id: &TokenIdentifier,
+        amount: &Self::BigUint,
+        function: &BoxedBytes,
+    ) -> AsyncCall<Self::SendApi> {
+        let contract_call: ContractCall<Self::SendApi, ()> =
+            ContractCall::new(self.send(), caller.clone(), function.clone())
+                .with_token_transfer(token_id.clone(), amount.clone());
+
+        contract_call
+            .async_call()
+            .with_callback(self.callbacks().async_transfer_execute_callback(caller))
     }
 
-    fn require_local_role_set(&self, token_id: &TokenIdentifier, role: &EsdtLocalRole) -> SCResult<()> {
+    fn require_local_role_set(
+        &self,
+        token_id: &TokenIdentifier,
+        role: &EsdtLocalRole,
+    ) -> SCResult<()> {
         let roles = self.blockchain().get_esdt_local_roles(token_id);
-        require!(
-            roles.contains(role),
-            "Must set local role first"
-        );
+        require!(roles.contains(role), "Must set local role first");
 
         Ok(())
     }
 
+    fn revert_operation_and_send(
+        &self,
+        address: &Address,
+        token_id: &TokenIdentifier,
+        amount: &Self::BigUint,
+    ) {
+        let wrapped_egld_token_id = self.wrapped_egld_token_id().get();
+
+        if token_id == &TokenIdentifier::egld() {
```

Comment:
Can use the `is_egld` method of `TokenIdentifier` instead of creating a separate instance to compare.

Reply:
Done